Source/WebCore/platform/graphics/ImageBuffer.cpp

```diff
@@ -32,25 +32,64 @@
 #include "GraphicsContext.h"
 #include "HostWindow.h"
 #include "PlatformImageBuffer.h"
+#include <wtf/text/Base64.h>
+
+#if USE(CG)
+#include "ImageBufferUtilitiesCG.h"
+#endif
+#if USE(CAIRO)
+#include "ImageBufferUtilitiesCairo.h"
+#endif
 
 namespace WebCore {
 
 static const float MaxClampedLength = 4096;
 static const float MaxClampedArea = MaxClampedLength * MaxClampedLength;
 
+static RefPtr<ImageBuffer> copyImageBuffer(Ref<ImageBuffer> source, PreserveResolution preserveResolution)
+{
+    if (source->resolutionScale() == 1 || preserveResolution == PreserveResolution::Yes) {
+        if (source->hasOneRef())
+            return source;
+    }
+    auto copySize = source->logicalSize();
+    auto copyScale = preserveResolution == PreserveResolution::Yes ? source->resolutionScale() : 1.f;
+    auto copyBuffer = source->context().createImageBuffer(copySize, copyScale, source->colorSpace());
+    if (!copyBuffer)
+        return nullptr;
+    if (source->hasOneRef())
+        ImageBuffer::drawConsuming(WTFMove(source), copyBuffer->context(), FloatRect { { }, copySize }, FloatRect { 0, 0, -1, -1 }, CompositeOperator::Copy);
+    else
+        copyBuffer->context().drawImageBuffer(source, FloatPoint { }, CompositeOperator::Copy);
+    return copyBuffer;
+}
+
+static RefPtr<NativeImage> copyNativeImageImpl(Ref<ImageBuffer> source, BackingStoreCopy copyBehavior, PreserveResolution preserveResolution)
+{
+    if (source->resolutionScale() == 1 || preserveResolution == PreserveResolution::Yes) {
+        if (source->hasOneRef())
```

Comment:
How can this be true? It is called like this:

`auto image = copyNativeImageImpl(const_cast<ImageBuffer&>(*this), copyBehavior, preserveResolution);`

This will construct a new Ref<ImageBuffer> and this will increment the refCount by 1. 

It is also called like this from `ImageBuffer::toDataURL()` and `ImageBuffer::toData()`

`image = copyNativeImageImpl(WTFMove(source), DontCopyBackingStore, preserveResolution);`

But the caller of these functions call them like this:

`return toDataURL(Ref { const_cast<ImageBuffer&>(*this) }, mimeType, quality, preserveResolution);`

So when this function is called the refCount is at least equal to 2.

Reply:
Yeah I'm not sure I understand this part either. Will ping Kimmo to add his thoughts here.